tools/autograd/templates/variable_factories.h

```diff
@@ -107,91 +142,57 @@ AT_FORALL_SCALAR_TYPES_AND3(Bool, Half, BFloat16, TENSOR)
     }
 
     void pretty_print_recursive(std::ostream& stream) const {
-      if (type_ == InitListTensorType::Scalar) {
-        AT_DISPATCH_ALL_TYPES_AND3(at::kBool, at::kHalf, at::kBFloat16, scalar_type_, "InitListTensor_pretty_print_scalar", [&] {
+      if (type_ == ListInitTensorType::Scalar) {
+        AT_DISPATCH_ALL_TYPES_AND3(at::kBool, at::kHalf, at::kBFloat16, scalar_type_, "ListInitTensor_pretty_print_scalar", [&] {
           stream << scalar_.to<scalar_t>();
         });
-      } else if (type_ == InitListTensorType::InitList) {
+      } else if (type_ == ListInitTensorType::Vector) {
         stream << "{";
-        for (const InitListTensor* it = init_list_.begin(); it != init_list_.end(); it++) {
+        for (auto it = vec_.begin(); it != vec_.end(); it++) {
           it->pretty_print_recursive(stream);
-          if (std::next(it) != init_list_.end()) stream << ", ";
+          if (std::next(it) != vec_.end()) stream << ", ";
         }
         stream << "}";
       }
     }
 
    private:
     void fill_tensor(at::Tensor tensor) const {
-      size_t index = 0;
-      for (const auto& elem : init_list_) {
-        if (elem.type_ == InitListTensorType::Scalar) {
-          at::NoGradGuard guard;
-          tensor[index].fill_(elem.scalar());
-        } else if (elem.type_ == InitListTensorType::InitList) {
+      if (type_ == ListInitTensorType::Scalar) {
```

Comment:
Why did this change?

Reply:
This is changed so that we can call `ListInitTensor.fill_tensor(scalar_tensor)` to fill a scalar tensor (the original code assumes that the tensor is not scalar, and goes into each slice of the tensor in order to fill them). After this change, `torch::tensor(3.14)` can work with `ListInitTensor`.

Reply:
So what I'm confused about is that `torch::tensor(3.14)` used to work, before this pull request, which means that this change wasn't necessary before but it is now. Why is it necessary now?

Reply:
Oh, I see now. All of the previous tensor ctors got deleted

Reply:
Synced offline - the rationale for the change is as follows:
1. Because `torch::tensor({{1, 2}})` is causing ambiguous overload currently (explained in the PR description), we need to remove the `torch::tensor(...)` overload that takes `c10::ArrayRef<T>`, and let `ListInitTensor` handle `c10::ArrayRef<T>` inputs as well.
2. `torch::tensor(scalar_value)` calls `torch::tensor(c10::ArrayRef<T>)`, and since we are removing the `torch::tensor(c10::ArrayRef<T>)` overload, we need to let `ListInitTensor` handle scalar inputs as well.
3. We make the changes here, so that `ListInitTensor` can handle scalar inputs.